src/qbox_statemachine_demo.py

```diff
@@ -0,0 +1,219 @@
+
+test1 = [("nodejs", QboxConfig.FAIL_404), ("ruby", QboxConfig.SUCCESS_200), ("ruby2", QboxConfig.FAIL_404)]
+test_index = 0
+def get_respond():
+    global test_index
+    res = test1[test_index]
+    test_index += 1
+    return res
+
+def send_compensation(config):
+    print("compensation sent")
+    return 1
+
+def send_back_result_to_requestor(config, status):
+    if status:
+        s = "failed"
+    else:
+        s = "succeed"
+    print("result is sent back to requestor, this time it is " + s)
+    return 1
+
+class StateInfo:
+    INIT = 1
+    AWAIT = 2
+    SUCCESS = 3
+    FAIL = 4
+    COMPENSATED = 5
+    PARTIALLY_FAIL = 6
+    ALL_FAIL = 7
+    ALL_SUCCESS = 8
+    def __init__(self, config, bitmap):
+        self.config = config
+        self.bitmap = bitmap
+    
+    def change_state(self, i, state):
+        self.bitmap[i] = state
+
+    def get_state(self, i):
+        return self.bitmap[i]
+
+    def get_all_response(self):
+        for bit in self.bitmap:
+            if bit == StateInfo.INIT or bit == StateInfo.AWAIT:
+                return False
+        return True
+
+    def get_response_status(self):
+        has_fail = False
+        has_success = False
+        for bit in self.bitmap:
+            if bit == StateInfo.FAIL:
+                if has_success:
+                    return StateInfo.PARTIALLY_FAIL
+                else:
+                    has_fail = True
+            elif bit == StateInfo.SUCCESS:
+                if has_fail:
+                    return StateInfo.PARTIALLY_FAIL
+                else:
+                    has_success = True
+            else:
+                assert(0)
+        if has_fail:
+            return StateInfo.ALL_FAIL
+        else:
+            return StateInfo.ALL_SUCCESS
+
+    def __repr__(self):
+        s = ''
+        for bit, transaction in zip(self.bitmap, config["onrequest"]["transactions"]):
+            s += "transaction: " + transaction["dstname"] + " state is: "
+            if bit == StateInfo.INIT:
+                s += "initial"
+            elif bit == StateInfo.AWAIT:
+                s += "awaiting response"
+            elif bit == StateInfo.SUCCESS:
+                s += "success"
+            elif bit == StateInfo.FAIL:
+                s += "fail"
+            elif bit == StateInfo.COMPENSATED:
+                s += "compensated"
+            s += "\n"
+        return s
+class State:
+    start_state = 1
+    awaiting_state = 2
+    compensation_state = 3
+    finish_state = 4
+
+class StateMachine:
+    def __init__(self, config):
+        self.config = config
+        self.transaction_number = len(config["onrequest"]["transactions"])
+        self.state_info = StateInfo(config, [StateInfo.INIT for i in range(self.transaction_number)])
+        self.state_type = State.start_state
+        self.response = [None] * self.transaction_number
+        self.saga_fail = False
+
+    def transition(self):
+        if self.state_type == State.start_state:

```

Comment:
So we need to connect this to the work happening in `qbox.py`. There, if a new request comes in, it should trigger a state change for an existing `StateMachine` or create a new `StateMachine` for that request type.

Look at the `handler` function in `qbox.py` and see if you can implement that integration there. Once that's done, I can approve this PR.

Reply:
To be clear, you can expect every request to have a header called `X-Qbox-RequestType` that can be `sendreq` or other values defined in your config. If the request has no header, you should simply send it forward to Envoy (this you can do by sending the body and headers of that request as a new request somewhere).